### PR TITLE
feat(breaker): ValueDeltaBreaker

### LINKED ISSUE #85: Static Value Breaker

### PR DESCRIPTION
**Pull request by bowd**
### Description

Value Delta Breaker is an extension of the Median Delta Breaker, but instead of moving with the median, this breaker ensures that the oracle value always stays in a threshold of a reference value. This is useful for the USDC/USD oracles where we want to ensure that trading stops if the reported USDC price moves outside of a comfortable range, 1 USD +/- 5bps for example.

### Other changes

Because breakers share a lot of logic when it comes to their cooldown and threshold management, I abstracted that away in utility contracts that are shared by both the MedianDeltaBreaker and the ValueDeltaBreaker.

### Tested

Copied over the MedianDeltaBreaker tests and changed them accordingly.

### Related issues

- Fixes #85 

### Backwards compatibility

Changes are additive

### Documentation

The BreakerBox doc need to be updated.

### CODE REVIEW

File: test/ValueDeltaBreaker.t.sol [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    assertEq(address(breaker.sortedOracles()), address(sortedOracles));
+  }
+
+  function test_constructor_shouldSetRateChangeThresholds() public {
+    assertEq(breaker.rateChangeThreshold(rateFeedIDs[0]), rateChangeThresholds[0]);
+  }
+
+  /* ---------- Setters ---------- */
+
+  function test_setCooldownTime_whenCallerIsNotOwner_shouldRevert() public {
```
Comment (bayological) [inline test/ValueDeltaBreaker.t.sol:93]: Test naming.
```suggestion
  function test_setDefaultCooldownTime_whenCallerIsNotOwner_shouldRevert() public {
```

File: test/ValueDeltaBreaker.t.sol [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+  }
+
+  function test_setRateChangeThreshold_whenCallerIsNotOwner_shouldRevert() public {
+    vm.expectRevert("Ownable: caller is not the owner");
+    changePrank(notDeployer);
+
+    breaker.setDefaultRateChangeThreshold(123456);
+  }
+
+  function test_setRateChangeThreshold_whenValueGreaterThanOne_shouldRevert() public {
```
Comment (bayological) [inline test/ValueDeltaBreaker.t.sol:116]: Test naming.
```suggestion
  function test_setDefaultRateChangeThreshold_whenValueGreaterThanOne_shouldRevert() public {
```

File: test/ValueDeltaBreaker.t.sol [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    vm.expectRevert("value must be less than 1");
+    breaker.setDefaultRateChangeThreshold(1 * 10**24);
+  }
+
+  function test_setRateChangeThreshold_whenCallerIsOwner_shouldUpdateAndEmit() public {
```
Comment (bayological) [inline test/ValueDeltaBreaker.t.sol:121]: Test naming.
```suggestion
  function test_setDefaultRateChangeThreshold_whenCallerIsOwner_shouldUpdateAndEmit() public {
```

File: contracts/ValueDeltaBreaker.sol [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+  /* ==================== Restricted Functions ==================== */
+
+  /**
+   * @notice Sets the cooldown time to the specified value for a rate feed.
+   * @param rateFeedIDs the targeted rate feed.
+   * @param cooldownTimes The new cooldownTime value.
+   * @dev Should be set to 0 to force a manual reset.
+   */
+  function setCooldownTimes(address[] calldata rateFeedIDs, uint256[] calldata cooldownTimes) external onlyOwner {
```
Comment (bayological) [inline contracts/ValueDeltaBreaker.sol:59]: We should add a test that verifies this function works as expected. Maybe consider adding tests for the logic of both WithCoolDown & WithThreshold to separate tests then in the individual breaker tests we can just verify access control works as intended?

File: test/ValueDeltaBreaker.t.sol [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+    address newSortedOracles = actor("newSortedOracles");
+    vm.expectEmit(true, true, true, true);
+    emit SortedOraclesUpdated(newSortedOracles);
+
+    breaker.setSortedOracles(ISortedOracles(newSortedOracles));
+
+    assertEq(address(breaker.sortedOracles()), newSortedOracles);
+  }
+
+  function test_setRateChangeThreshold_whenSenderIsNotOwner_shouldRevert() public {
```
Comment (bayological) [inline test/ValueDeltaBreaker.t.sol:152]: Typo.
```suggestion
  function test_setRateChangeThresholds_whenSenderIsNotOwner_shouldRevert() public {
```

File: test/ValueDeltaBreaker.t.sol [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+    changePrank(notDeployer);
+    vm.expectRevert("Ownable: caller is not the owner");
+    breaker.setRateChangeThresholds(rateFeedIDs, rateChangeThresholds);
+  }
+
+  function test_setRateChangeThreshold_whenValuesAreDifferentLengths_shouldRevert() public {
```
Comment (bayological) [inline test/ValueDeltaBreaker.t.sol:158]: Typo.
```suggestion
  function test_setRateChangeThresholds_whenValuesAreDifferentLengths_shouldRevert() public {
```

File: test/ValueDeltaBreaker.t.sol [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+    address[] memory rateFeedIDs2 = new address[](2);
+    rateFeedIDs2[0] = actor("randomRateFeed");
+    rateFeedIDs2[1] = actor("randomRateFeed2");
+    vm.expectRevert("array length missmatch");
+    breaker.setRateChangeThresholds(rateFeedIDs2, rateChangeThresholds);
+  }
+
+  function test_setRateChangeThreshold_whenThresholdIsMoreThan0_shouldRevert() public {
```
Comment (bayological) [inline test/ValueDeltaBreaker.t.sol:166]: Typo.
```suggestion
  function test_setRateChangeThresholds_whenThresholdIsMoreThan0_shouldRevert() public {
```

File: test/ValueDeltaBreaker.t.sol [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+    rateChangeThresholds[0] = 1 * 10**24;
+    vm.expectRevert("value must be less than 1");
+    breaker.setRateChangeThresholds(rateFeedIDs, rateChangeThresholds);
+  }
+
+  function test_setRateChangeThreshold_whenSenderIsOwner_shouldUpdateAndEmit() public {
```
Comment (bayological) [inline test/ValueDeltaBreaker.t.sol:172]:
```suggestion
  function test_setRateChangeThresholds_whenSenderIsOwner_shouldUpdateAndEmit() public {
```

File: contracts/MedianDeltaBreaker.sol [hunk 9]
```diff
@@ -1,7 +1,9 @@
-// SPDX-License-Identifier: UNLICENSED
+// SPDX-Liaense-Identifier: UNLICENSED
```
Comment (bayological) [inline contracts/MedianDeltaBreaker.sol:1]:
```suggestion
// SPDX-License-Identifier: UNLICENSED
```

File: contracts/MedianDeltaBreaker.sol [hunk 10]
```diff
@@ -16,91 +18,70 @@ import { FixidityLib } from "./common/FixidityLib.sol";
...
     ISortedOracles _sortedOracles
   ) public {
     _transferOwnership(msg.sender);
-    setCooldownTime(_cooldownTime);
-    setDefaultRateChangeThreshold(_defaultRateChangeThreshold);
+    _setDefaultCooldownTime(_defaultCooldownTime);
+    _setDefaultRateChangeThreshold(_defaultRateChangeThreshold);
     setSortedOracles(_sortedOracles);
-    setRateChangeThresholds(rateFeedIDs, rateChangeThresholds);
+    // setRateChangeThresholds(rateFeedIDs, rateChangeThresholds);
```
Comment (bayological) [inline contracts/MedianDeltaBreaker.sol:43]: Assuming this is intentional, we should remove the comment. But also, why not initialise the thresholds on deployment?

File: test/MedianDeltaBreaker.t.sol [hunk 11]
```diff
@@ -198,28 +175,27 @@ contract MedianDeltaBreakerTest_constructorAndSetters is MedianDeltaBreakerTest
...
+    assertEq(breaker.getCooldown(rateFeedIDs[0]), coolDownTime);
   }
 }
 
 contract MedianDeltaBreakerTest_shouldTrigger is MedianDeltaBreakerTest {
   function updateMedianByPercent(uint256 medianChangeScaleFactor, address _rateFeedID) public {
     uint256 previousMedianRate = 0.98 * 10**24;
     uint256 currentMedianRate = (previousMedianRate * medianChangeScaleFactor) / 10**24;
-    setupSortedOracles(currentMedianRate, previousMedianRate);
+    stdstore.target(address(breaker)).sig(0x7dffc308).with_key(_rateFeedID).checked_write(previousMedianRate);
```
Comment (bayological) [inline test/MedianDeltaBreaker.t.sol:186]: Nice.🧙‍♂️

**Review by bayological — CHANGES_REQUESTED** [review]

File: contracts/common/breakers/WithCooldown.sol [hunk 12]
```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.5.13;
+
+/**
+ * @title   Breaker With Cooldown
+ * @notice  Utility portion of a Breaker contract which deals with the
+ *          cooldown component.
+ */
+contract WithCooldown {
```
[the remark this answers is not in the document]
Comment (bowd) [reply contracts/common/breakers/WithCooldown.sol:9]: I wish I could but I can't.

**Review by bayological — APPROVED** [review]
LGTM

**Pull request merged**